### PR TITLE
Multi vcf download plugin

### PR DESCRIPTION
**Pull request by skjnldsv**
@charismatic-claire @brjhaverkamp

This plugin allow us to ~~POST~~GET a list of vcards as a multi dimensional array and retrieve everything as a single vcf file. Needed for https://github.com/nextcloud/contacts/issues/39

To discuss:
1. ~~shall we go with a GET instead?~~ DONE
2. ~~shall we use a simple array instead of a multi dimensional one?~~ DONE

``` bash
curl 'https://dev.skjnldsv.com/remote.php/dav/addressbooks/users/admin/contacts/?export' \
	-X REPORT  --user 'admin:admin' \
	--data-binary '<x1:addressbook-multiget xmlns:x1="urn:ietf:params:xml:ns:carddav"><x0:prop xmlns:x0="DAV:"><x0:getcontenttype/><x0:getetag/><x0:resourcetype/><x0:displayname/><x0:owner/><x0:resourcetype/><x0:sync-token/><x0:current-user-privilege-set/><x0:getcontenttype/><x0:getetag/><x0:resourcetype/><x1:address-data/></x0:prop><x0:href xmlns:x0="DAV:">83F18885-8FF6-4839-9D10-C997D5BD9B96.vcf</x0:href><x0:href xmlns:x0="DAV:">470623A2-0FCB-4BA0-A6D2-6C25F4113C02.vcf</x0:href></x1:addressbook-multiget>'
```

### CODE REVIEW

File: apps/dav/lib/CardDAV/VCFMultiExportPlugin.php [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	/** @var Server */
+	protected $server;
+
+	/**
+	 * Initializes the plugin and registers event handlers
+	 *
+	 * @param Server $server
+	 * @return void
+	 */
+	function initialize(Server $server) {
```
Comment (ChristophWurst) [inline apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]: public? private? protected?

**Review by ChristophWurst — DISMISSED** [review]

File: apps/dav/lib/CardDAV/VCFMultiExportPlugin.php (hunk 1 above)
Comment (skjnldsv) [reply apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]: :see_no_evil: :bowing_man:

**Comment by ChristophWurst** [comment]
>     1. shall we go with a GET instead?

If that call does not change any server state, yes.

**Comment by skjnldsv** [comment]
> > ```
> > 1. shall we go with a GET instead?
> > ```
> 
> If that call does not change any server state, yes.

Yes, I'm thinking that as well, but since muti dimentionnal arrays are terribly looking as url, I wonder if some browsers can go :boom: with that?
`/remote.php/dav/addressbooks/users/admin/?vcards%5Bcontacts%5D%5B%5D=3A6B15C7-EDCF-4F14-B9DB-D427806B82A2.vcf&vcards%5Btest%5D%5B%5D=790AD673-6C80-47F9-95A7-89E9AD21B801.vcf`

File: apps/dav/appinfo/v1/carddav.php [hunk 2]
```diff
@@ -84,6 +84,7 @@
 
 $server->addPlugin(new \Sabre\DAV\Sync\Plugin());
 $server->addPlugin(new \Sabre\CardDAV\VCFExportPlugin());
+$server->addPlugin(new \OCA\DAV\CardDAV\VCFMultiExportPlugin());
```
Comment (rullzer) [inline apps/dav/appinfo/v1/carddav.php]: I would be fine not adding this to the legacy endpoint

**Comment by rullzer** [comment]
Also you hit the various browser url length limits at some point

File: apps/dav/appinfo/v1/carddav.php (hunk 2 above)
Comment (skjnldsv) [reply apps/dav/appinfo/v1/carddav.php]: I was wondering as well! :)

**Comment by skjnldsv** [comment]
> Also you hit the various browser url length limits at some point

Arf, damn!
What should I do then? :(

File: apps/dav/lib/CardDAV/VCFMultiExportPlugin.php [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+		foreach ($nodes as $node) {
+
+			$nodeData = $node->get();
+
+			// Parsing this node so VObject can clean up the output.
+			$vcard = VObject\Reader::read($nodeData);
+			$output .= $vcard->serialize();
+
+			// Destroy circular references to PHP will GC the object.
```
Comment (ChristophWurst) [inline apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]:
```suggestion
			// Destroy circular references so PHP will GC the object.
```

File: apps/dav/lib/CardDAV/VCFMultiExportPlugin.php [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	/**
+	 * Merges all vcard objects, and builds one big vcf export
+	 *
+	 * @param array $nodes
+	 * @return string
+	 */
+	private function generateVCF(array $nodes) {
+		$output = '';
+
+		foreach ($nodes as $node) {
```
Comment (ChristophWurst) [inline apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]: FYI: `array_reduce` would be handy here to reduce the `$nodes` to a string like

```php
return array_reduce($nodes, function($carry, $node) {
     return $carry . /* your magic */;
}, '');
```

File: apps/dav/lib/CardDAV/VCFMultiExportPlugin.php [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+			$node = $this->server->tree->getNodeForPath("$path/$addressbook/");
+
+			// Checking ACL, if available.
+			if ($aclPlugin = $this->server->getPlugin('acl')) {
+				$aclPlugin->checkPrivileges("$path/$addressbook/", '{DAV:}read');
+			}
+
+			// checking that the path is a valid addressbook
+			if ($node instanceof AddressBook) {
+				$addressbooks = array_merge($addressbooks, [$addressbook => $vcards]);
```
Comment (ChristophWurst) [inline apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]:
```suggestion
				$addressbooks[$addressbook] = $vcards;
```

would do the exact same I guess

**Review by ChristophWurst — COMMENTED** [review]
Some general coding style feedback. Can't argue much about the logic as I'm not familiar with the details of DAV and VCF :man_shrugging:

File: apps/dav/lib/CardDAV/VCFMultiExportPlugin.php (hunk 4 above)
Comment (skjnldsv) [reply apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]: Yes, I used that at first, but ended up using the core function made by the creator. :)

File: apps/dav/lib/CardDAV/VCFMultiExportPlugin.php (hunk 5 above)
Comment (skjnldsv) [reply apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]: Actually I can drop this I think now. I don't use $addressbooks anymore! :+1:

File: apps/dav/lib/CardDAV/VCFMultiExportPlugin.php (hunk 4 above)
Comment (ChristophWurst) [reply apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]: what core function?

Comment (skjnldsv) [reply apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]: https://github.com/nextcloud/3rdparty/blob/master/sabre/dav/lib/CardDAV/VCFExportPlugin.php#L113

Comment (ChristophWurst) [reply apps/dav/lib/CardDAV/VCFMultiExportPlugin.php]: "code reuse"

**Comment by rullzer** [comment]
> > Also you hit the various browser url length limits at some point
> 
> Arf, damn!
> What should I do then? :(

Well using POST helps for that. With GET you could run into isues if people select their whole addressbook

**Comment by MorrisJobke** [comment]
> > 
> 
> Hum, no better ideas out there? Anyone?

This is the only way I also have in mind :/

**Comment by ChristophWurst** [comment]
> Also you hit the various browser url length limits at some point

Found https://stackoverflow.com/a/2659995/2239067 and https://stackoverflow.com/a/417184/2239067 telling me the limit is at ~2KB. Based on the example URL provided I assume a typical URL parameter is up to 100 bytes longs, allowing just 20 contacts/parameters.

Since this is DAV anyway, this answer might be relevant: https://stackoverflow.com/questions/978061/http-get-with-request-body#comment10368662_978061

IMO using POST for this seems wrong as it's an idempotent operation. Still, GET is also not right, esp. with the technical limitations.

**Comment by charismatic-claire** [comment]
How about the REPORT request, that was specifically invented for that kind of a request against a CardDAV server? See [Building a CardDAV client](http://sabre.io/dav/building-a-carddav-client/), search for the keyword "multiget".

Or is that impossible, because we are limited to GET, POST, PUT and DELETE here?

**Comment by georgehrke** [comment]
@skjnldsv I don't think we need a dedicated plugin for that. You can perform a simple [addressbook-multiget](https://github.com/nextcloud/cdav-library/blob/fee4d2f4b4c0345507544404c338d2b97747228b/src/models/addressBook.js#L181), put them all together into one big vcf in javascript and use a data-uri link to download.

**Comment by georgehrke** [comment]
> How about the REPORT request, that was specifically invented for that kind of a request against a CardDAV server? See Building a CardDAV client, search for the keyword "multiget".

I agree that multiget is the way to go, but please take note that multiget does not return one big vcf file, but a WebDav like 207-multistatus response with individual vcfs. See https://tools.ietf.org/html/rfc6352#section-8.7.1 :)

**Comment by skjnldsv** [comment]
@georgehrke @charismatic-claire 
So I would then build the vcf in the browser?
That's something I really don't like. If I have a selection of 1000 vcfs, I want to just trigger a download. Any javascript computation will explode the browser, especially if the vcfs are big :(

**Comment by charismatic-claire** [comment]
@skjnldsv 
I think the whole point of building that plugin server side was to not let the browser do the heavy lifting. But I see your point. If the Contacts app fires a multiget request via REPORT and receives lots of single VCFs as a response, it will have to do exactly that... Hm... Can't our plugin just return one single VCF as a result? But that would be against the standard, right? Hm... So the question becomes: Is there a standard for our use case, or do we have to invent one? And if so, is that a good idea? Do I miss something here?

**Comment by georgehrke** [comment]
Well, you could adapt the VCFExportPlugin to allow appending `?export` to the url when sending `REPORT` requests.

**Comment by skjnldsv** [comment]
> Well, you could adapt the VCFExportPlugin to allow appending `?export` to the url when sending `REPORT` requests.

then build a blob from the report request and trigger a redirect to the blob?
Could be clean yes!!

@charismatic-claire you're right on the point yes!
Thank you both!! I will adapt this later.

@charismatic-claire could you then adapt this behavior to the contacts app? :)

**Comment by charismatic-claire** [comment]
@skjnldsv I'll be glad to help :-)

**Comment by skjnldsv** [comment]
@georgehrke done! :)
Should you implement this into the cdavlib then?

**Comment by skjnldsv** [comment]
@MorrisJobke @rullzer @ChristophWurst a little review for the php?

**Comment by ChristophWurst** [comment]
Is this plugin even needed now? The discussion above – esp https://github.com/nextcloud/server/pull/12004#issuecomment-433001723 and https://github.com/nextcloud/server/pull/12004#issuecomment-433128384 – tells me this might not be required if I understand correctly.

**Comment by skjnldsv** [comment]
@ChristophWurst I updated the plugin to use the multi-get like @georgehrke said :)

If you do a multiget with ?export at the end, it will return a vcf file and not a dav response ;)

``` bash
curl 'https://dev.skjnldsv.com/remote.php/dav/addressbooks/users/admin/contacts/?export' \
	-X REPORT  --user 'admin:admin' \
	--data-binary '<x1:addressbook-multiget xmlns:x1="urn:ietf:params:xml:ns:carddav"><x0:prop xmlns:x0="DAV:"><x0:getcontenttype/><x0:getetag/><x0:resourcetype/><x0:displayname/><x0:owner/><x0:resourcetype/><x0:sync-token/><x0:current-user-privilege-set/><x0:getcontenttype/><x0:getetag/><x0:resourcetype/><x1:address-data/></x0:prop><x0:href xmlns:x0="DAV:">83F18885-8FF6-4839-9D10-C997D5BD9B96.vcf</x0:href><x0:href xmlns:x0="DAV:">470623A2-0FCB-4BA0-A6D2-6C25F4113C02.vcf</x0:href></x1:addressbook-multiget>'
```

**Comment by charismatic-claire** [comment]
@ChristophWurst Maybe every pull request should start with a comprehensive description of the "ticket" we are trying to solve here. I'll try to briefly summarize what we wanted to achieve:

- In the `Contacts` app, we want to add the possibility to export whole groups of contacts. Say you have 5 address books and 10 groups. What you might want is to export all members from all address books belonging to a certain group. Or maybe you want to search for a certain phrase and want to export all contacts matching that search. Or you want to manually select a number of contacts you want to export.
- By "export" in mean you want to download one single vcf file containing all the vCards of all the contacts you want to export.
- Here is what we wanna do: To achieve this, the interface of the `Contacts` app shall compile a `REPORT` request for all the wanted contacts. If you send this to the CardDAV server, it will respond with a list of all the contacts. So the transformation into one single vcf file which could be downloaded would be done on the browser side via JavaScript. This is what @skjnldsv wants to avoid here.
- That's why we need this plugin here. With a `export` parameter appended to the URL of the `REPORT` request, this plugin would transform this response into one single vcf file.
- See [Currently not possible to download group of contacts](https://github.com/nextcloud/contacts/issues/39) for more

Does that make clear what we want and why we need this plugin on the server side? Feel free to ask, if not :-)

**Review by georgehrke — DISMISSED** [review]
And please squash all commits :)

File: apps/dav/lib/CardDAV/MultiGetExportPlugin.php [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+		// Reduce the vcards into one string
+		$output = array_reduce($responseXml->getResponses(), function ($vcf, $card) {
+			$vcf .= $card->getResponseProperties()[200]['{urn:ietf:params:xml:ns:carddav}address-data'];
+			return $vcf;
+		}, '');
+
+		// Build and override the response
+		$filename = 'vcfexport-' . date('Y-m-d') . $filenameExtension;
+		$response->setHeader('Content-Disposition', 'attachment; filename="' . $filename . '"');
+		$response->setHeader('Content-Type', 'text/directory');
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply apps/dav/lib/CardDAV/MultiGetExportPlugin.php]: > This should be `text/vcard`

This is not what sabre/dav is doing https://github.com/nextcloud/3rdparty/blob/master/sabre/dav/lib/CardDAV/VCFExportPlugin.php#L83 :woman_shrugging: 

Okay for 200! :+1: :wink:

**Comment by ChristophWurst** [comment]
Thank you, @charismatic-claire :v:

**Review by georgehrke — APPROVED** [review]

**Review by ChristophWurst — APPROVED** [review]
Fine, I guess :elephant:

**Comment by skjnldsv** [comment]
Failure unrelated

**Pull request merged**